### PR TITLE
Encapsulate socket in client

### PR DESCRIPTION
**Pull request by jden**
References
========
- [RADAR-602](https://zendesk.atlassian.net/browse/RADAR-602)

Required
========
- [ ] :+1: from @zendesk/zendesk-radar

Risks
========
- Medium: clientSession.send now expects an objects rather than a JSON string, and clientSession.on('message', handler) now emits JSON-parsed objects. The raw engineio sockets send and emit JSON strings. Internal references have been updated; other code which was previously interacting directly with engineio sockets will have to be updated to match this new signature.
- Medium: adds client session state model. Messages won't be emitted from clientSession once it has ended.
- Low: change in how client sessions are looked up by ids. Other code which is directly accessing sockets on the engineio socketServer instance will have to be updated to instead get sessions from the ClientSession module. In the future, this should be cleaned up further.

### CODE REVIEW

**Comment by jden** [comment]
This pulls an abstract `transport` into the `ClientSession` class. Next step is to find all the references to `socket` and replace them with `ClientSession` and use the `message` event and the `send` method on the `ClientSession` instance instead.

File: client/client_session.js [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+  message.to = messageIn.to
+  message.op = messageIn.op
+
+  return message
+}
+
+// TODO: move these to SessionManager
+var clientsById = {}
+var ClientIdsByName = {} // keyed by name
+var ClientNamesById = {} // keyed by id
```
Comment (vanchi-zendesk) [inline client/client_session.js]: maybe hoist these up to the top?

File: client/client_session.js [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+  if (message.op !== 'nameSync') { return }
+
+  log.info('#socket.message - nameSync', message, this.id)
+
+  this.send({ op: 'ack', value: message && message.ack })
+
+  var association = message.options.association
+
+  ClientNamesById[association.id] = association.name
+  ClientIdsByName[association.name] = this
```
Comment (vanchi-zendesk) [inline client/client_session.js]: Just call them ClientNames and Clients.

File: server/server.js [hunk 3]
```diff
@@ -73,80 +73,48 @@ Server.prototype.terminate = function (done) {
...
-
-    this.engineioPath = configuration.engineio.conf ? configuration.engineio.conf.path : 'default'
-  }
-
   this._setupSentry(configuration)
-
-  this.socketServer = engine.attach(httpServer, engineConf)
-  this.socketServer.on('connection', this._onSocketConnection.bind(this))
+  this._setupEngineio(httpServer, configuration.engineio)
+  this._setupDistributor()
```
Comment (vanchi-zendesk) [inline server/server.js]: some of these rewrites are not very related to clientSession, although they are good rewrites. I suggest we split the PRs up (just do a  separate PR into this branch, or master and then rebase) so that reviewing is a bit easier.

File: client/client_session.js [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+  }
+
+  if (this.state.can('initialize')) {
+    this.state.initialize()
+  }
+  this.lastModified = Date.now()
+}
+
+ClientSession.prototype._cleanup = function () {
+  delete clientsById[this.id]
```
Comment (vanchi-zendesk) [inline client/client_session.js]: should we not cleanup the name as well?

**Comment by vanchi-zendesk** [comment]
two concerns: Logging might be too much, so may want to change to debug on the session logging. Another thing is fix memory leaks when clients disconnect within the session hashes. 

Otherwise it looks good. I will look at the fix commits, so please don't rebase.

File: client/client_session.js (hunk 2 above)
Comment (jden) [inline client/client_session.js]: The values in ClientIdsByName are strings (the ids), not ClientSession objects.

File: server/server.js (hunk 3 above)
Comment (jden) [inline server/server.js]: Broken out into https://github.com/zendesk/radar/pull/199

**Comment by jden** [comment]
@vanchi-zendesk 
1) the logging in `client_session.js` is not new, it replaces the logging for `#socket.message.outgoing` log messages. It should not increase log volume
2) I removed storing the name/id association maps, since it was not used (so now no leak). The name is stored on the ClientSession object in `nameSync` message handler. Per our chat, we may need to update this. There is no logic change in this PR from what was happening previously on master- places that are now calling `ClientSession.get()` (using the id) were previously doing lookups based on the engineio `socket.id`

File: client/client_session.js (hunk 4 above)
Comment (jden) [inline client/client_session.js]: I removed the name associations, see https://github.com/zendesk/radar/pull/198#issuecomment-168849148

**Comment by vanchi-zendesk** [comment]
:+1:
